### PR TITLE
Fix noob'sturn

### PR DESCRIPTION
**Pull request by pedrorribeiro**

### CODE REVIEW

File: main.py [hunk 1]
```diff
@@ -119,9 +129,23 @@ def draw_whitespace(drawing: Drawing):
...
+
+start = 0
+while start == 0:
+    start = input('Ready to start the window? (1 - yes, 0 - no)')
+
 pygame.init()
 SCREEN = pygame.display.set_mode((800, 600))
 
+position_right = 0
+while position_right == 0:
```
Comment (pedrorribeiro) [inline main.py:207]: this doesn't work and I don't know why

File: main.py [hunk 2]
```diff
@@ -119,9 +129,23 @@ def draw_whitespace(drawing: Drawing):
 shapes = load_shapes()
 positions = ["top_left", "top_right", "bottom_left", "bottom_right"]
 
+subject_ID = input('Insert Subject ID')
+subject = Subject(subject_ID, date.today(), {})
+subject.create_data_template()
+
+
+start = 0
+while start == 0:
```
Comment (pedrorribeiro) [inline main.py:200]: this doesn't work and I don't know why

File: .idea/.name [hunk 3]
```diff
@@ -0,0 +1 @@
+subject.py
```
Comment (Gevie) [inline .idea/.name:1]: We should add `.idea/` to `.gitignore` as these are files generated by your IDE.

File: main.py [hunk 4]
```diff
@@ -8,20 +8,95 @@
...
 os.environ["SDL_VIDEO_CENTERED"] = "1"
 
+path = './IAPS/*.jpg'
+
+IAPS = []
+for filename in glob.glob(path):
+    IAPS.append(filename)
+
+
+def get_iap(file):
```
Comment (Gevie) [inline main.py:27]: I'm not sure why you needed to write this, are you just trying to get the name of the image? There are many handy ways to do this, https://www.pythonpool.com/python-get-filename-without-extension/

Failing that, you could put this as a for loop so you're not repeating your code like this

File: main.py [hunk 5]
```diff
@@ -97,18 +172,12 @@ def draw(drawing: Drawing, single_shape: bool = False) -> Shape:
...
-    path = './IAPS/*.jpg'
-
-    IAPS = []
-    for filename in glob.glob(path):
-        IAPS.append(filename)
-
     random.shuffle(IAPS)
     figure = IAPS.pop()
-
+    plus = get_iap(figure)
```
Comment (Gevie) [inline main.py:177]: I don't really know what plus means, I would consider naming it something more descriptive.

File: main.py (hunk 2 above)
Comment (Gevie) [reply main.py:200]: Maybe just do

```python
while True:
    start_game = input('Ready to start the game? (Type Y to start):\n')
    if start_game == 'Y':
        break
```

File: main.py (hunk 1 above)
Comment (Gevie) [reply main.py:207]: Like above, something like this should work

```python
while True:
    position_check = input('Is the game positioned correctly? (Type Y to confirm):\n')
    if position_check == 'Y':
        break
        
    # Quit the application? Restart the process? etc
```

**Review by Gevie — COMMENTED** [review]
I'm going to finish my review here because I'm not seeing a full scope of changes. We're building pending PR on top of pending PR on top of pending PR - I will decline the last two and point this one to master.

File: subject.py [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+                self.date: {
+                    'Timestamps': {
+                        'Baseline': {},
+                        'Stimulation': {}
+                    }
+                }
+            }
+        }
+
+    def add_timestamp(self, current_round: Round, words, stage, timestamp: str) -> None:
```
Comment (Gevie) [inline subject.py:90]: With this one, we don't need to pass `stage` because we already have the `Round` which we can use to get the current stage.

We also shouldn't need to pass `words` (which isn't very descriptive), because if we are keeping `Round.current_round` and `Round.iap` up to date we can just use those instead.

File: subject.py [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+        Stores the timestamp for a round against a round
+
+        Args:
+            current_round (Round): The round object
+            timestamp (str): The timestamp to store against said round
+
+        Returns:
+            None
+        """
+        if stage == 'Game':
```
Comment (Gevie) [inline subject.py:101]: I would say this could be greatly simplified like such

**Version 1**
```python
game_stages = ['Game', 'SingleChoiceGame']
if current_round.stage in game_stages:
    round_name = f'Q{current_round.current_round}'
    self.data[self.id][self.date]['Timestamps']['Baseline'].update({round_name: timestamp})
    return None

round_name = f'IAPS{current_round.iap}' if current_round.stage == 'Stimulation' else f'Baseline{current_round.iap}'
self.data[self.id][self.date]['Timestamps']['Stimulation'].update({round_name: timestamp})
```

If you really brushed up how Round and Instance were communicating, you could just use the if statements to set the round name and then add it to data

**Version 2** (Better in my opinion)
```python
# Be a good idea to make these class properties
game_stages = ['Game', 'SingleChoiceGame']
round_prefixes = {
    'Stimulation': 'IAPS',
    'Baseline': 'Baseline',
    'Game': 'Q',
    'SingleChoiceGame': 'Q'
}

prefix = round_prefixes[current_round.stage]
round_number = current_round.current_round if current.round.stage in game_stages else current_round.iap
round_name = f'{prefix}{round_number}'
    
self.data[self.id][self.date]['Timestamps'][current_round.stage].update({round_name: timestamp})
```

It's also worth nothing that `pass` isn't needed if you're doing an `elif`.

File: main.py [hunk 8]
```diff
@@ -6,106 +6,277 @@
...
+Instance.add_timestamp(current_round, stage)
 
 while True:
     for events in pygame.event.get():
         if events.type == pygame.QUIT:
             pygame.quit()
             sys.exit()
 
-        if events.type == generate_new_round and current_round <= 24:
+        if events.type == generate_new_round and current_round <= 81:
```
Comment (Gevie) [inline main.py:230]: We built a class to represent a round and an instance class to handle logic with incrementation and switching stage etc, somehow we've gone back to the 81 round count iterator and tracking rounds separately here.

It is a bit of a mess and hard to follow, but since you need to run this experiment for real tomorrow and it's now working for you, just leave it as it is. (For now)

You should also consider using elif for the current round checks here.

**Review by Gevie — COMMENTED** [review]
I've left some more feedback, you don't need to action any of this since you are running this experiment tomorrow and I'd hate for you to break your now working version before that happens. (Although in git, you could always checkout the last commit which was working so it's not all too bad if that happens).

I'd like to work on this after you've run your experiment, make it something which anybody could use with their own parameters and rules - decouple it from being tied so directly to your experiment requirements.

File: subject.py (hunk 6 above)
Comment (pedrorribeiro) [reply subject.py:90]: There is a weird bug that occurs if you do not pass, where it will load all 40 whitespace rounds at once and all 20 image rounds right after, ignoring the timer. Could not find reference to why that is

File: main.py (hunk 8 above)
Comment (pedrorribeiro) [reply main.py:230]: If we're going to use this later on, we need to remake the round logic from the ground up. The subject save function was also broken, but I changed it to work for my own devices. The problem was syntax for the json class, I believe

Comment (pedrorribeiro) [reply main.py:230]: I am planning on working on this again on Thursday as well, when I don't have a time limit looming over me, cause I know a lot of that could be better, but I took 8 hours to make this one work, so I need a wee break XD

File: subject.py (hunk 6 above)
Comment (pedrorribeiro) [reply subject.py:90]: It happens wayyyyy less frequently with the pass structure

File: subject.py (hunk 7 above)
Comment (pedrorribeiro) [reply subject.py:101]: This seems good, if we can make the Round.stage logic work again it should be good. Unfortunately I could not do that

Comment (Gevie) [reply subject.py:101]: Actually, we could improve this greatly by separating a bit of logic out, then the class method would just be

```python
    def add_timestamp(self, current_round: Round, timestamp: str) -> None:
        """
        Stores the timestamp for a round against a round

        Args:
            current_round (Round): The round object
            timestamp (str): The timestamp to store against said round

        Returns:
            None
        """
        
        round_name = current_round.get_round_name()
        self.data[self.id][self.date]['Timestamps'][current_round.stage].update({round_name: timestamp})
```

Comment (pedrorribeiro) [reply subject.py:101]: I can write the new get_round name, if you want, since I wrote the logic for the IAPS number and the rest is straight forward

Comment (Gevie) [reply subject.py:101]: Sure thing, I probably won't be able to contribute code updates until the week after next. I've got a lot to do at home, places to be, and a wedding to go away for.

Comment (pedrorribeiro) [reply subject.py:101]: Of course, you have said that you were unavailable

File: subject.py (hunk 6 above)
Comment (Gevie) [reply subject.py:90]: That will likely be happening outside the method, best fix Thursday onwards is just to trace what is really happening and fixing the problem at the root cause (when you're not on a deadline)

File: main.py (hunk 8 above)
Comment (Gevie) [reply main.py:230]: Absolutely, take all the time you need, you don't owe me anything I'm the one assisting you with your project / learning. 😅 This does seem like a good project for you to come back to and collaborate on to improve your skills so I'd be glad to help you whip this into better shape and slowly introduce better design patterns to replace the old stuff.
